### PR TITLE
feat(eslint-plugin): [restrict-template-expressions] add support for intersection types

### LINKED ISSUE #1797: [restrict-template-expressions] Missing handling of intersection type
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/restrict-template-expressions": "error"
  }
}
```

```TS
type MyString = string & { kind: "MyString" };

const myString = "x" as MyString;

console.log(`myString: ${myString}`); // False positive "Invalid type of template ..."
```

**Expected Result**

No error.

**Actual Result**

"Invalid type of template literal expression"

**Additional Info**

Real-world example:

```ts
import { Term } from "sparqljs";

const someTerm = "x" as Term;

console.log(`someTerm: ${someTerm}`);
```

See https://github.com/DefinitelyTyped/DefinitelyTyped/blob/638a579eecfa0486115f2440518603f25a8b58b7/types/sparqljs/index.d.ts#L223.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.25.0` |
| `@typescript-eslint/parser`        | `2.25.0` |

**Proposed solution**

Don't emit an error if _any_ item of the intersection type is "allowed" in respect to the options.

### PR DESCRIPTION
**Pull request by ulrichb**
Fixes #1797

@reviewers: I had to refactor the `getBaseType()` function (which extracted all _union_ primitive/"other" types) into a predicate-receiving `isUnderlyingExpressionTypeConfirmingTo()`.

IMO this is better anyways because it separates the two concerns: 
a) "a property x applies to all underlying types", and 
b) "some type x is primitive (in respect to the options)"

Also `isUnderlyingExpressionTypeConfirmingTo()` is more generic (easier to re-use).

Better ideas for the name "isUnderlyingExpressionTypeConfirmingTo"?

### CODE REVIEW

**Comment by ulrichb** [comment]
@bradzacher Also adapted the commit message (fix=>feat)

File: packages/eslint-plugin/src/rules/restrict-template-expressions.ts [hunk 1]
```diff
@@ -76,75 +83,52 @@ export default util.createRule<Options, MessageId>({
...
+      expression: TSESTree.Expression,
+      predicate: (underlyingType: ts.Type) => boolean,
+    ): boolean {
+      const expressionType = getExpressionNodeType(expression);
 
-      return getBaseType(type);
-    }
+      return rec(
+        // "Extracts" generic constraint, indexed access and conditional types:
+        typeChecker.getBaseConstraintOfType(expressionType) ?? expressionType,
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/restrict-template-expressions.ts:110]: We've got a utility function for this!
```suggestion
        util.getConstrainedTypeAtLocation(typeChecker, expression),
```

File: packages/eslint-plugin/src/rules/restrict-template-expressions.ts [hunk 2]
```diff
@@ -44,29 +44,36 @@ export default util.createRule<Options, MessageId>({
...
+      if (util.isTypeFlagSet(type, ts.TypeFlags.StringLike)) {
+        return true;
+      }
+
+      if (
+        util.isTypeFlagSet(
+          type,
+          ts.TypeFlags.NumberLike | ts.TypeFlags.BigIntLike,
+        ) &&
+        options.allowNumber
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/restrict-template-expressions.ts:57]: nit - do the option check first to save some cycles
```suggestion
        options.allowNumber &&
        util.isTypeFlagSet(
          type,
          ts.TypeFlags.NumberLike | ts.TypeFlags.BigIntLike,
        )
```

File: packages/eslint-plugin/src/rules/restrict-template-expressions.ts [hunk 3]
```diff
@@ -44,29 +44,36 @@ export default util.createRule<Options, MessageId>({
...
+      ) {
+        return true;
       }
-      return true;
+
+      if (
+        util.isTypeFlagSet(type, ts.TypeFlags.BooleanLike) &&
+        options.allowBoolean
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/restrict-template-expressions.ts:64]: ditto

```suggestion
        options.allowBoolean &&
        util.isTypeFlagSet(type, ts.TypeFlags.BooleanLike)
```

File: packages/eslint-plugin/src/rules/restrict-template-expressions.ts [hunk 4]
```diff
@@ -44,29 +44,36 @@ export default util.createRule<Options, MessageId>({
...
+      ) {
+        return true;
+      }
+
+      if (
+        util.isTypeFlagSet(type, ts.TypeFlags.Null | ts.TypeFlags.Undefined) &&
+        options.allowNullable
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/restrict-template-expressions.ts:71]: ditto
```suggestion
        options.allowNullable &&
        util.isTypeFlagSet(type, ts.TypeFlags.Null | ts.TypeFlags.Undefined)
```

**Comment by ulrichb** [comment]
@bradzacher 

* Merged master and added "allowAny" from d44c0f9bed2404ca00b020b35fd825929e213398
* Refactor to `util.getConstrainedTypeAtLocation()`
* Move the option-checks before the `isTypeFlagSet()`-checks
* Add example: `stringWithKind: string & { kind?: 'MyString' }`

**Comment by ulrichb** [comment]
Thanks for reviewing!

**Pull request merged**